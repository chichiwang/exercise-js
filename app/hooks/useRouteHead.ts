import { useLocation, useMatches } from 'react-router-dom';
import type { PageMatch, RouteHead } from 'app/types';

function useRouteHead(): RouteHead {
  const location = useLocation();
  const routeMatches = useMatches();

  const locationMatch = routeMatches.find(
    function matchLocationWithHead(match): match is PageMatch {
      const pathMatches = match.pathname === location.pathname;
      const containsHandle = typeof match.handle === 'object';
      const containsHead = containsHandle
        && Object.prototype.hasOwnProperty.call(match.handle, 'head');

      return pathMatches && containsHead;
    },
  );

  const pathChainMatches = routeMatches.filter(
    function matchesWithHead(match): match is PageMatch {
      const containsHandle = typeof match.handle === 'object';

      return containsHandle && Object.prototype.hasOwnProperty.call(match.handle, 'head');
    },
  );

  const pathChainMatch = pathChainMatches.length
    ? pathChainMatches[pathChainMatches.length - 1]
    : undefined;

  return locationMatch?.handle?.head || pathChainMatch?.handle?.head || {};
}

export default useRouteHead;
