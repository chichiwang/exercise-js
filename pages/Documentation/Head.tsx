/* global PROJECT_URL */
import React from 'react';

function Head() {
  const url = `${PROJECT_URL}/documentation`;

  return (
    <>
      <meta property="og:type" content="article" />
      <meta property="og:author" content="Chi-chi Wang" />
      <meta property="og:title" content="Developer Documentation for Tamsui: React/Express universal JavaScript boilerplate" />
      <meta property="og:description" content="Set up a universal JavaScript application on the Tamsui boilerplate." />
      <meta property="og:url" content={url} />
      <link href={url} rel="canonical" />
    </>
  );
}

export default Head;
