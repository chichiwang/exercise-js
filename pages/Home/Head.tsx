/* global PROJECT_URL */
import React from 'react';

function Head() {
  const url = `${PROJECT_URL}`;

  return (
    <>
      <meta property="og:type" content="website" />
      <meta property="og:title" content="Tamsui: A React/Express universal JavaScript boilerplate" />
      <meta property="og:description" content="Build a universal JavaScript application using React and Express." />
      <meta property="og:url" content={url} />
    </>
  );
}

export default Head;
