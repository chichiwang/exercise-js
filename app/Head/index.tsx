/* global PROJECT_URL */
import React, { useContext } from 'react';

import ManifestContext from 'app/context/ManifestContext';
import useRouteHead from 'app/hooks/useRouteHead';
import type { Manifest } from 'app/types';

function Head() {
  const manifest: Manifest = useContext(ManifestContext);
  const head = useRouteHead();

  const defaultTitle = 'Tamsui: A Universal JavaScript Application Boilerplate';
  const titleSuffix = '| Tamsui';
  const title = head?.title
    ? `${head.title} ${titleSuffix}`
    : defaultTitle;

  return (
    <head>
      <meta charSet="utf-8" />
      <title>{ title }</title>
      <meta name="author" content="Chi-chi Wang" />
      <meta
        name="keywords"
        content="Tamsui, js, JavaScript, isomorphic, Node.js, React, TypeScript, Express, React-Router, Boilerplate"
      />
      <meta name="description" content="An Express/React universal application boilerplate" />
      <meta name="robots" content="all" />
      <meta name="viewport" content="width=device-width" />
      <link rel="stylesheet" type="text/css" href={`/${manifest['app.css']}`} />
      <link rel="apple-touch-icon" sizes="180x180" href="/static/images/favicon/apple-touch-icon.png" />
      <link rel="icon" type="image/png" sizes="32x32" href="/static/images/favicon/favicon-32x32.png" />
      <link rel="icon" type="image/png" sizes="16x16" href="/static/images/favicon/favicon-16x16.png" />
      <link rel="manifest" href="/static/site.webmanifest" />
      <meta property="og:locale" content="en_US" />
      <meta property="og:image" content={`${PROJECT_URL}/static/images/site-preview.webp`} />
      <meta property="og:image:width" content="1200" />
      <meta property="og:image:height" content="630" />
      <meta property="og:image:type" content="image/webp" />
    </head>
  );
}

export default Head;
