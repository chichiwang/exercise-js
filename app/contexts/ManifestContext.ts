import { createContext } from 'react';

import { Manifest } from 'app/types';

const ManifestContext = createContext<Manifest>({});

export default ManifestContext;
