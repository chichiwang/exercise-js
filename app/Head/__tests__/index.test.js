import React from 'react';
import renderer from 'react-test-renderer';

import ManifestContext from 'app/context/ManifestContext';
import useRouteHead from 'app/hooks/useRouteHead';

import Head from '../index';

jest.mock('app/hooks/useRouteHead', () => jest.fn());

const mockedManifest = {
  'app.css': 'Head CSS filepath',
};

describe('Head component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('matches the no route head snapshot', () => {
    useRouteHead.mockReturnValueOnce({});

    const tree = renderer.create(
      <ManifestContext.Provider value={mockedManifest}>
        <Head />
      </ManifestContext.Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('matches the route title snapshot', () => {
    useRouteHead.mockReturnValueOnce({
      title: 'route has PageHandle with title',
    });

    const tree = renderer.create(
      <ManifestContext.Provider value={mockedManifest}>
        <Head />
      </ManifestContext.Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
