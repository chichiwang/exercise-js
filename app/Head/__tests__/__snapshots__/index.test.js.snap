// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Head component matches the no route head snapshot 1`] = `
<head>
  <meta
    charSet="utf-8"
  />
  <title>
    Tamsui: A Universal JavaScript Application Boilerplate
  </title>
  <meta
    content="Chi-chi Wang"
    name="author"
  />
  <meta
    content="Tamsui, js, JavaScript, isomorphic, Node.js, React, TypeScript, Express, React-Router, Boilerplate"
    name="keywords"
  />
  <meta
    content="An Express/React universal application boilerplate"
    name="description"
  />
  <meta
    content="all"
    name="robots"
  />
  <meta
    content="width=device-width"
    name="viewport"
  />
  <link
    href="/Head CSS filepath"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="/static/images/favicon/apple-touch-icon.png"
    rel="apple-touch-icon"
    sizes="180x180"
  />
  <link
    href="/static/images/favicon/favicon-32x32.png"
    rel="icon"
    sizes="32x32"
    type="image/png"
  />
  <link
    href="/static/images/favicon/favicon-16x16.png"
    rel="icon"
    sizes="16x16"
    type="image/png"
  />
  <link
    href="/static/site.webmanifest"
    rel="manifest"
  />
</head>
`;

exports[`Head component matches the route title snapshot 1`] = `
<head>
  <meta
    charSet="utf-8"
  />
  <title>
    route has PageHandle with title | Tamsui
  </title>
  <meta
    content="Chi-chi Wang"
    name="author"
  />
  <meta
    content="Tamsui, js, JavaScript, isomorphic, Node.js, React, TypeScript, Express, React-Router, Boilerplate"
    name="keywords"
  />
  <meta
    content="An Express/React universal application boilerplate"
    name="description"
  />
  <meta
    content="all"
    name="robots"
  />
  <meta
    content="width=device-width"
    name="viewport"
  />
  <link
    href="/Head CSS filepath"
    rel="stylesheet"
    type="text/css"
  />
  <link
    href="/static/images/favicon/apple-touch-icon.png"
    rel="apple-touch-icon"
    sizes="180x180"
  />
  <link
    href="/static/images/favicon/favicon-32x32.png"
    rel="icon"
    sizes="32x32"
    type="image/png"
  />
  <link
    href="/static/images/favicon/favicon-16x16.png"
    rel="icon"
    sizes="16x16"
    type="image/png"
  />
  <link
    href="/static/site.webmanifest"
    rel="manifest"
  />
</head>
`;
